### PR TITLE
Fix: Handle logging/setLevel request for server

### LINKED ISSUE #293: logging/setLevel request not handled
**Describe the bug**

After initialization, MCP clients (such as the MCP inspector) will send a `logging/setLevel` request if the logging capability is enabled.  The logging capability is enabled by default, but this request is not handled, leading to an error that blocks the use of the MCP server (at least in the inspector).

**To Reproduce**
1.  Create a SSE server with the default `ServerCapabilities` object
2. Try to connect to it in the MCP inspector

**Expected behavior**

The logging level request is handled, and sets the log level for the server.

**DoD**

- [ ] Handiling of `logging/setLevel` is implemented
- [ ] Integration test

### PR DESCRIPTION
**Pull request by devcrocod**
Fix #293 

Also fixed a bug with initialization: properties must be declared before the init block; otherwise, the initialization order is violated and an NPE will occur in the init block.

## Motivation and Context
After initialization, MCP clients send a `logging/setLevel` request if the logging capability is enabled. This request was previously not handled, leading to an error that blocked the use of the MCP server in the inspector and other clients.

## How Has This Been Tested?
Added integration test

## Breaking Changes
None

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

### CODE REVIEW

File: kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt [hunk 1]
```diff
@@ -71,6 +56,44 @@ public open class ServerSession(
...
+    /**
+     * The capabilities supported by the server, related to the session.
+     */
+    private val serverCapabilities = options.capabilities
+
+    /**
+     * The current logging level set by the client.
+     * When null, all messages are sent (no filtering).
+     */
+    private var currentLoggingLevel: LoggingLevel? = null
```
[the remark this answers is not in the document]
Comment (kpavlov) [reply kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt]: true

File: kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt [hunk 2]
```diff
@@ -152,12 +175,20 @@ public open class ServerSession(
...
+     * Messages are filtered based on the current logging level set by the client.
+     * If no logging level is set, all messages are sent.
      *
      * @param notification The logging message notification.
      */
     public suspend fun sendLoggingMessage(notification: LoggingMessageNotification) {
-        logger.trace { "Sending logging message: ${notification.params.data}" }
-        notification(notification)
+        if (serverCapabilities.logging != null) {
+            if (!isMessageIgnored(notification.params.level)) {
```
Comment (kpavlov) [inline kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt]: Consider avoiding negation in `if` conditions, as it can make the code harder to read. For instance, using `if (isMessageAccepted(...)) { ... }` is more straightforward than a negated condition.

File: kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt [hunk 3]
```diff
@@ -71,6 +56,44 @@ public open class ServerSession(
...
+     * When null, all messages are sent (no filtering).
+     */
+    private var currentLoggingLevel: LoggingLevel? = null
+
+    /**
+     * Map of LoggingLevel to severity index for comparison.
+     * Higher index means higher severity.
+     */
+    private val loggingLevelSeverity: Map<LoggingLevel, Int> = LoggingLevel.entries.withIndex()
+        .associate { (index, level) -> level to index }
```
Comment (kpavlov) [inline kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt]: This can be a constant; there is no reason to calculate and store it as an instance field. Should be a part of the `LoggingLevel` enum.

BTW, why not use `LoggingLevel#ordinal`?

File: kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt [hunk 4]
```diff
@@ -152,12 +175,20 @@ public open class ServerSession(
 
     /**
      * Sends a logging message notification to the client.
+     * Messages are filtered based on the current logging level set by the client.
+     * If no logging level is set, all messages are sent.
```
Comment (kpavlov) [inline kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt:182]: I'm not sure if any message should be sent if no level is explicitly defined.

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt [hunk 5]
```diff
@@ -886,6 +886,96 @@ class ClientTest {
...
+        )
+
+        serverSession.sendLoggingMessage(
+            LoggingMessageNotification(
+                params = LoggingMessageNotification.Params(
+                    level = LoggingLevel.error,
+                    data = buildJsonObject { put("message", "Error - should pass") },
+                ),
+            ),
+        )
```
Comment (kpavlov) [inline kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt]: nit: can be replaced with a loop

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt [hunk 6]
```diff
@@ -886,6 +886,96 @@ class ClientTest {
...
+
+        // Only warning and error should be received
+        assertEquals(2, receivedMessages.size, "Should receive only 2 messages (warning and error)")
```
Comment (kpavlov) [inline kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt:957]: Shouldn't [awaitility](https://github.com/awaitility/awaitility) be used here?

It will be needed for sure to verify [rate-limiting](https://modelcontextprotocol.io/specification/2025-06-18/server/utilities/logging#implementation-considerations)

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt [hunk 7]
```diff
@@ -886,6 +886,96 @@ class ClientTest {
...
+
+        val levels = receivedMessages.map { it.params.level }
+        assertTrue(levels.contains(LoggingLevel.warning), "Should receive warning message")
+        assertTrue(levels.contains(LoggingLevel.error), "Should receive error message")
```
[the remark this answers is not in the document]
Comment (kpavlov) [reply kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt]: true

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt [hunk 8]
```diff
@@ -886,6 +886,96 @@ class ClientTest {
...
+            launch {
+                serverSessionResult.complete(server.connect(serverTransport))
+                println("Server connected")
+            },
+        ).joinAll()
+
+        val serverSession = serverSessionResult.await()
+
+        // Set logging level to warning
+        val result = client.setLoggingLevel(LoggingLevel.warning)
```
Comment (kpavlov) [inline kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt]: It's easy to write a parametrized test verifying all logging levels

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt (hunk 7 above)
Comment (kpavlov) [inline kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt]: can be replaced with a loop verifying that `levels` contain all the messages with severity >= defined LoggingLevel's severity

**Review by kpavlov — CHANGES_REQUESTED** [review]
Let's make `currentLoggingLevel` thread-safe and use `LoggingLevel#ordinal` instead of calculating the severity level in a loop for every ServerSession instance, and improve the test.

[Consider edge cases](https://modelcontextprotocol.io/specification/2025-06-18/server/utilities/logging)

File: kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt (hunk 4 above)
Comment (devcrocod) [reply kotlin-sdk-server/src/commonMain/kotlin/io/modelcontextprotocol/kotlin/sdk/server/ServerSession.kt:182]: Yes, it should

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt (hunk 6 above)
Comment (devcrocod) [reply kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt:957]: it's common code, I used `delay`

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt [hunk 9]
```diff
@@ -886,6 +886,86 @@ class ClientTest {
...
+                LoggingMessageNotification(
+                    params = LoggingMessageNotification.Params(
+                        level = level,
+                        data = buildJsonObject { put("message", message) },
+                    ),
+                ),
+            )
+        }
+
+        delay(100)
```
Comment (kpavlov) [inline kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt:954]: awaitility is more correct than delay

**Review by kpavlov — APPROVED** [review]
Thank you, @devcrocod

File: kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt (hunk 9 above)
Comment (devcrocod) [reply kotlin-sdk-test/src/commonTest/kotlin/io/modelcontextprotocol/kotlin/sdk/client/ClientTest.kt:954]: > it's common code, I used delay
